numba/_dictobject.c

```diff
@@ -0,0 +1,1079 @@
+
+/* Alternative fraction that is otherwise close enough to 2n/3 to make
+ * little difference. 8 * 2/3 == 8 * 5/8 == 5. 16 * 2/3 == 16 * 5/8 == 10.
+ * 32 * 2/3 = 21, 32 * 5/8 = 20.
+ * Its advantage is that it is faster to compute on machines with slow division.
+ * #define USABLE_FRACTION(n) (((n) >> 1) + ((n) >> 2) - ((n) >> 3))
+ */
+
+/* GROWTH_RATE. Growth rate upon hitting maximum load.
+ * Currently set to used*3.
+ * This means that dicts double in size when growing without deletions,
+ * but have more head room when the number of deletions is on a par with the
+ * number of insertions.  See also bpo-17563 and bpo-33205.
+ *
+ * GROWTH_RATE was set to used*4 up to version 3.2.
+ * GROWTH_RATE was set to used*2 in version 3.3.0
+ * GROWTH_RATE was set to used*2 + capacity/2 in 3.4.0-3.6.0.
+ */
+#define GROWTH_RATE(d) ((d)->ma_used*3)
+
+void
+numba_dictkeys_free(NB_DictKeys *dk) {
+    free(dk);
+}
+
+void
+numba_dict_free(NB_Dict *d) {
+    numba_dictkeys_free(d->keys);
+    free(d);
+}
+
+Py_ssize_t
+numba_dict_length(NB_Dict *d) {
+    return d->used;
+}
+
+
+static NB_DictEntry*
+get_entry(NB_DictKeys *dk, Py_ssize_t idx) {
+    Py_ssize_t offset;
+    char *ptr;
+
+    assert (idx < dk->size);
+    offset = idx * dk->entry_size;
+    ptr = dk->indices + dk->entry_offset + offset;
+    return (NB_DictEntry*)ptr;
+}
+
+static void
+zero_key(NB_DictKeys *dk, char *data){
+    memset(data, 0, dk->key_size);
+}
+
+static void
+zero_val(NB_DictKeys *dk, char *data){
+    memset(data, 0, dk->val_size);
+}
+
+static void
+copy_key(NB_DictKeys *dk, char *dst, const char *src){
+    memcpy(dst, src, dk->key_size);
+}
+
+static void
+copy_val(NB_DictKeys *dk, char *dst, const char *src){
+    memcpy(dst, src, dk->val_size);
+}
+
+/* Returns -1 for error; 0 for not equal; 1 for equal */
+static int
+key_equal(NB_DictKeys *dk, const char *lhs, const char *rhs) {
+    /* XXX: should use user provided Equality */
+    return memcmp(lhs, rhs, dk->key_size) == 0;
+}
+
+static char *
+entry_get_key(NB_DictKeys *dk, NB_DictEntry* entry) {
+    char * out = entry->keyvalue;
+    assert (out == palign(out));
+    return out;
+}
+
+static char *
+entry_get_val(NB_DictKeys *dk, NB_DictEntry* entry) {
+    char * out = entry_get_key(dk, entry) + align(dk->key_size);
+    assert (out == palign(out));
+    return out;
+}
+
+/* Allocate new dictionary keys
+
+Adapted from CPython's new_keys_object().
+*/
+int
+numba_dictkeys_new(NB_DictKeys **out, Py_ssize_t size, Py_ssize_t key_size, Py_ssize_t val_size) {
+    Py_ssize_t usable = USABLE_FRACTION(size);
+    Py_ssize_t index_size = ix_size(size);
+    Py_ssize_t entry_size = align(sizeof(NB_DictEntry) + align(key_size) + align(val_size));
+    Py_ssize_t entry_offset = align(index_size * size);
+    Py_ssize_t alloc_size = sizeof(NB_DictKeys) + entry_offset + entry_size * usable;
+
+    NB_DictKeys *dk = palign(malloc(align(alloc_size)));
+    if (!dk) return ERR_NO_MEMORY;
```

Comment:
I'm going to remove the `palign` to `dk`.  I think it's better to assume malloc returning word size aligned pointer and that we are not freeing modified pointers.

Reply:
I did check the maths manually for a couple of arrays of allocation size and assumed pointer alignments coming from `malloc` and nothing immediately jumped out as invalid or weird, but if it's unnecessary then removing it will likely be safer.